odata-protocol/11.4 Data Modification.md

```diff
@@ -654,49 +668,72 @@ operations.
 On failure, the service MUST NOT apply any of the changes specified in
 the request.
 
-### ##subsubsec Upsert an Entity
+#### ##subsubsubsec Upsert an Entity
 
-An upsert occurs when the client sends an [update
-request](#UpdateanEntity) to a valid URL that identifies a single entity
-that does not yet exist. In this case the service MUST handle the
-request as a [create entity request](#CreateanEntity) or fail the
-request altogether.
+Services MAY treat an upsert request like a [create entity request](#CreateanEntity)
+if it requires the addressed entity to be newly created. The request is then said to be
+"treated as an insert", otherwise "treated as an update".
+Services that support this capability (instead of failing upsert requests that require
+entity creation) SHOULD advertise it by an annotation with the
+term `Capabilities.UpdateRestrictions` (nested property `Upsertable`
+with value `true`) defined in [OData-VocCap](#ODataVocCap).
 
-Upserts to single-valued navigation properties are possible for
+If the service is unable to determine the
+canonical collection or canonical singleton (as defined in [section ##ContextURL]) for the newly-created entity,
+it MUST fail the upsert request.
+This determination is possible, for example, if the resource path of the request
+1. consists of an entity set followed by a key predicate
+2. consists of a singleton (which must be nullable if the request can be treated as an insert)
+3. ends with a containment navigation property,
+   followed by a key predicate if this is collection-valued
+   (and which must be nullable otherwise if the request can be treated as an insert)
```

Comment:
I would not try to cover both single- and collection-valued in one bullet; it makes it too confusing (especially with the part about nullability).

Reply:
We could combine entity set and collection-value containment navigation property, and combine singleton and single-valued containment navigation property.